sympy/polys/domains/algebraicfield.py

```diff
@@ -569,6 +562,59 @@ def galois_group(self, by_name=False, max_tries=30, randomize=False):
             by_name=by_name, max_tries=max_tries, randomize=randomize)
 
 
+def _make_input_converter(K):
+    """Construct the converter to convert from Expr to K"""
+    # to_number_field is expensive, so we will try to reduce the number of
+    # calls to it by reducing down to the basic algebraic elements and then
+    # cache the actual calls to to_number_field. E.g 2*sqrt(2) + 3*sqrt(3) will
+    # be reduced to sqrt(2) and sqrt(3) and then to_number_field will be called
+    # on those using the cached converter.
+    #
+    # XXX: Most of the time the calls to isomorphism_converter are unnecessary.
+    # They result from e.g. QQ.algebraic_field(sqrt(2), sqrt(3)) trying to
+    # convert sqrt(2) and sqrt(3) to the field. When primitive_element is
+    # called it computes the conversion for the original elements but by this
+    # point we have forgotten that conversion...
```

Comment:
Do you suppose it would be worth caching the conversion?

The field's primitive element is computed here:

https://github.com/sympy/sympy/blob/ea1e3852851a9a84f124dea874307fa9ae11ab95/sympy/polys/domains/algebraicfield.py#L293-L302

with a call to `to_number_field()`, which is sort of a Swiss Army Knife function that ultimately uses `primitive_element()`. The field constructor could instead call `primitive_element()` directly, and could then opt to pass `ex=True`, in order to obtain the desired representations.

Currently `ex=False` is used, meaning the representations aren't just *forgotten* by now; they were never actually computed. I wonder if computing them is significantly slower, or if it would be worth obtaining them at construction time.

Reply:
I'm not sure. The code in `primitive_element` is here:
https://github.com/sympy/sympy/blob/ea1e3852851a9a84f124dea874307fa9ae11ab95/sympy/polys/numberfields/subfield.py#L350-L388
I would have guessed that the expensive part is `_factor_list` and potentially `_choose_factor` (if it needs to numerically evaluate RootOf).

The extra pieces to generate the representation of the elements seem like they would be cheap in the grand scheme of things. It is difficult to measure though because so many things are cached so these timings may not be representative:
```python
In [1]: surds = lambda n: [sqrt(prime(i)) for i in range(1, n+1)]

In [2]: %time ok = primitive_element(surds(2))
CPU times: user 106 ms, sys: 0 ns, total: 106 ms
Wall time: 105 ms

In [3]: %time ok = primitive_element(surds(3))
CPU times: user 67.2 ms, sys: 3.94 ms, total: 71.1 ms
Wall time: 70 ms

In [4]: %time ok = primitive_element(surds(4))
CPU times: user 82.7 ms, sys: 3.6 ms, total: 86.3 ms
Wall time: 84.8 ms

In [5]: %time ok = primitive_element(surds(5))
CPU times: user 147 ms, sys: 0 ns, total: 147 ms
Wall time: 145 ms

In [6]: %time ok = primitive_element(surds(6))
CPU times: user 315 ms, sys: 132 μs, total: 315 ms
Wall time: 317 ms
```
Compared to:
```python
In [1]: surds = lambda n: [sqrt(prime(i)) for i in range(1, n+1)]

In [2]: %time ok = primitive_element(surds(2), ex=True)
CPU times: user 181 ms, sys: 11.5 ms, total: 193 ms
Wall time: 191 ms

In [3]: %time ok = primitive_element(surds(3), ex=True)
CPU times: user 53.4 ms, sys: 87 μs, total: 53.4 ms
Wall time: 52.2 ms

In [4]: %time ok = primitive_element(surds(4), ex=True)
CPU times: user 88.6 ms, sys: 73 μs, total: 88.7 ms
Wall time: 87.7 ms

In [5]: %time ok = primitive_element(surds(5), ex=True)
CPU times: user 141 ms, sys: 3.77 ms, total: 144 ms
Wall time: 144 ms

In [6]: %time ok = primitive_element(surds(6), ex=True)
CPU times: user 423 ms, sys: 4.26 ms, total: 427 ms
Wall time: 429 ms
```
In general in SymPy is there ever a situation where we call `primitive_element` but would not want `ex=True`?

It seems like it is never going to be sufficient just to have the primitive element if you don't also have a way to represent the components of the primitive element in the number field.

Reply:
It's hard to say. The `ex=False` default was already in there when I refactored `polys.numberfields` a few years ago; i.e. it was not my design choice.

I agree though, it does seem reasonable that, if you're starting with multiple generators for a field, then you're going to want to continue to work with those generators, once you have the field.

For example, the fact that the `AlgebraicField` class stores the original generators under `self.orig_ext` attests to their continued importance.

But it's hard to guess whether the proposed change represents an overall performance penalty or boost for the average user.

Bottom line: I think this PR is fine as is. But if you wanted to try caching the generator representations at build time, I think I'd support that too (unless someone comes in with a great argument against it).

Reply:
Most of the time all the elements are converted to the domain at the same after the primitive element is computed inside construct domain which uses `ex=True`:
https://github.com/sympy/sympy/blob/96836db06ba6b78103dd4217db53db31502fb2f6/sympy/polys/constructor.py#L105-L128
So actually we already used `ex=True` before the AlgebraicField was even constructed but we didn't pass that information to the AlgebraicField constructor.

In the case in the OP I construct the AlgebraicField and only later ask to convert lots of expressions into it. I think it probably does make sense to use `ex=True` and store the results if a primitive element needs to be constructed in the AlgebraicField constructor.

At the same time though there can be other cases where we have expressions that are not literally part of the primitive element expression but still fit in the domain e.g.:
```python
In [9]: K = QQ.algebraic_field(exp(I*pi/7))

In [10]: K
Out[10]: QQ<exp(I*pi/7)>

In [11]: a = K.from_sympy(cos(2*pi/7)-I*sin(2*pi/7))

In [12]: a
Out[12]: ANP([-1, 0, 0, 0, 0, 0], [1, -1, 1, -1, 1, -1, 1], QQ)

In [13]: K.to_sympy(a)
Out[13]: 
  5⋅ⅈ⋅π
  ─────
    7  
-ℯ  
```
So I think it is still useful to cache the results of `to_number_field` in `K.from_sympy` either way. Actually I just found a bug for this case in this PR...